File: jput-core/src/main/java/cz/encircled/jput/unit/UnitPerformanceAnalyzerImpl.java
package cz.encircled.jput.unit;

import cz.encircled.jput.JPutContext;
import cz.encircled.jput.Statistics;
import cz.encircled.jput.model.MethodConfiguration;
import cz.encircled.jput.model.PerformanceTestRun;

import java.lang.reflect.Method;

/**
 * @author Vlad on 21-May-17.
 */
public class UnitPerformanceAnalyzerImpl implements UnitPerformanceAnalyzer {

    @Override
    public PerformanceTestRun buildRun(MethodConfiguration configuration, Method method) {
        PerformanceTestRun run = new PerformanceTestRun();
        run.runs = new long[configuration.repeats];
        run.testMethod = method.getName();
        run.testClass = method.getDeclaringClass().getName();
        run.executionId = JPutContext.getContext().getContextExecutionId();
        return run;
    }

    private String toString(PerformanceTestRun run, MethodConfiguration configuration) {
        String stats;
        stats = String.format("[ averageTime = %d ms, maxTime = %d ms]", Statistics.averageRunTime(run), Statistics.maxRunTime(run));
        return "PerformanceTestRun{" +
                "testMethod = " + run.testMethod +
                ", configuration = " + configuration +
                ", run = " + stats +
                '}';
    }

    @Override
    public void addRun(PerformanceTestRun run, long elapsedTime) {
        run.runs[run.positionCounter++] = elapsedTime;
    }

    @Override
    public UnitPerformanceResult analyzeUnitTrend(PerformanceTestRun run, MethodConfiguration conf) {
        UnitPerformanceResult result = new UnitPerformanceResult();

        long runAvgTime = Statistics.averageRunTime(run);
        if (conf.averageTimeLimit > 0 && runAvgTime > conf.averageTimeLimit) {
            result.isAverageLimitMet = false;
            result.runAverageTime = runAvgTime;
        }

        long runMaxTime = Statistics.maxRunTime(run);
        if (conf.maxTimeLimit > 0 && runMaxTime > conf.maxTimeLimit) {
            result.isMaxLimitMet = false;
            result.runMaxTime = runMaxTime;
        }
//                for (Map.Entry<Long, Long> percentile : conf.percentiles.entrySet()) { TODO
//                    long matchingCount = LongStream.of(run.runs).filter(time -> time <= percentile.getValue()).count();
//                    int matchingPercents = Math.round(matchingCount * 100 / conf.repeats);
//                    if (matchingPercents < percentile.getKey()) {
//                        String assertMessage = "\nMax time = " + percentile.getValue() + "ms \nexpected percentile = " + percentile.getKey() +
//                                "%\nActual percentile = " + matchingPercents + "%\n\n";
//                        throw new AssertionFailedError(assertMessage + "Performance test failed, max time is greater then limit: " + analyzer.toString(run, conf));
//                    }
//                }

        return result;
    }

    @Override
    public String buildErrorMessage(UnitPerformanceResult result, MethodConfiguration conf) {
        String msg = "";
        if (!result.isAverageLimitMet) {
            msg += String.format("\nLimit avg time = %d ms\nActual avg time = %d ms", conf.averageTimeLimit, result.runAverageTime);
        }
        if (!result.isMaxLimitMet) {
            if (!result.isAverageLimitMet) {
                msg += "\n";
            }
            msg += String.format("\nLimit max time = %d ms\nActual max time = %d ms", conf.maxTimeLimit, result.runMaxTime);
        }
        return msg;
    }

}

File: jput-core/src/main/java/cz/encircled/jput/unit/UnitPerformanceAnalyzerImpl.kt
package cz.encircled.jput.unit

import cz.encircled.jput.model.PerfConstraintViolation
import cz.encircled.jput.model.PerfTestExecution

/**
 * @author Vlad on 21-May-17.
 */
interface UnitPerformanceAnalyzer {

    fun analyzeUnitTrend(execution: PerfTestExecution): List<PerfConstraintViolation>

}


/**
 * This analyzer verifies that execution times (avg, max etc) are within defined limits.
 *
 * @author Vlad on 21-May-17.
 */
class UnitPerformanceAnalyzerImpl : UnitPerformanceAnalyzer {

    override fun analyzeUnitTrend(execution: PerfTestExecution): List<PerfConstraintViolation> {
        val result = mutableListOf<PerfConstraintViolation>()
        if (execution.conf.avgTimeLimit > 0 && execution.executionAvg > execution.conf.avgTimeLimit) {
            result.add(PerfConstraintViolation.UNIT_AVG)
        }

        if (execution.conf.maxTimeLimit > 0 && execution.executionMax > execution.conf.maxTimeLimit) {
            result.add(PerfConstraintViolation.UNIT_MAX)
        }

        //  for (Map.Entry<Long, Long> percentile : conf.percentiles.entrySet()) { TODO
        //      long matchingCount = LongStream.of(execution.runs).filter(time -> time <= percentile.getValue()).count();
        //      int matchingPercents = Math.round(matchingCount * 100 / conf.repeats);
        //      if (matchingPercents < percentile.getKey()) {
        //          String assertMessage = "\nMax time = " + percentile.getValue() + "ms \nexpected percentile = " + percentile.getKey() +
        //                  "%\nActual percentile = " + matchingPercents + "%\n\n";
        //          throw new AssertionFailedError(assertMessage + "Performance test failed, max time is greater then limit: " + analyzer.toString(execution, conf));
        //      }
        //  }

        return result
    }

}

/**
 * This analyzer verifies that count of occurred exceptions is within defined limits
 */
class TestExceptionsAnalyzer : UnitPerformanceAnalyzer {

    override fun analyzeUnitTrend(execution: PerfTestExecution): List<PerfConstraintViolation> =
            if (execution.exceptionsCount() > execution.conf.maxAllowedExceptionsCount) {
                listOf(PerfConstraintViolation.EXCEPTIONS_COUNT)
            } else emptyList()

}